### PR TITLE
Use the right encoding to open files.

### LINKED ISSUE #3: grin3 assumes a wrong encoding for my shell
I have a shell in UTF-8 with files encoded in UTF-8.

However when using grin3, it seems to parse them in latin-1.

```bash
$ grin "Information d'intégration"
$ grin "Information d'int"
./dashboard/templates/admin/summary.html.j2:
   83 :         <h3 class="panel-title">Information d'intÃ©gration</h3>
```

### PR DESCRIPTION
**Pull request by Natim**
Fixes #3

### CODE REVIEW

**Comment by rsalmaso** [comment]
Hi, thanks.
Could you add an `--encoding`, `-x` option which defaults to `sys.stdout.encoding` and use this value?

File: grin/main.py [hunk 1]
```diff
@@ -92,15 +92,21 @@ def read_block_with_context(self, prev, fp, fp_size):
...
+            else:
+                remaining = max(fp_size - fp.tell(), 0)
+                target_io_size = min(READ_BLOCKSIZE, remaining)
+                block_main = fp.read(target_io_size)
+                is_last_block = target_io_size == remaining
+        except UnicodeDecodeError as e:  # Decode error
+            # Fallback to latin1
             target_io_size = READ_BLOCKSIZE
-            block_main = fp.read(target_io_size)
+            block_main = e.object.decode("latin1")
```
Comment (Natim) [inline grin/main.py]: I don't know how to handle this case to be honest. Is it particular to my setup?

Shall we have that to a list of possible encodings and try each of them until it doesn't fails?

File: grin/grin.py [hunk 2]
```diff
@@ -377,9 +386,11 @@ def main(argv=None):
         )
         args.use_color = args.force_color or use_term_color
 
+        encoding = args.encoding
+
```
Comment (rsalmaso) [inline grin/grin.py]:
```suggestion
```

File: grin/grin.py [hunk 3]
```diff
@@ -377,9 +386,11 @@ def main(argv=None):
...
         regex = get_regex(args)
         g = GrepText(regex, args)
-        openers = dict(text=open, gzip=gzip.open)
+        openers = dict(text=partial(codecs.open, encoding=encoding), gzip=gzip.open)
```
Comment (rsalmaso) [inline grin/grin.py]:
```suggestion
        openers = dict(text=partial(codecs.open, encoding=args.encoding), gzip=gzip.open)
```

**Review by rsalmaso — CHANGES_REQUESTED** [review]
I'm pretty ok with this. Could you add a test for this new option?

**Comment by Natim** [comment]
> Could you add a test for this new option?

I added tests about encoding management.

**Comment by rsalmaso** [comment]
Manually merged in https://github.com/rsalmaso/grin3/commit/51fdac178e02a5034f8cafaeeffad6cce340bb9e
Release grin3 2.2.0

**Pull request closed without merging**